File: python/examples/3.Webcam/run_client.py
import cv2
import time
from ezpysocket import ezpysocket as ps


def client_operation(c: ps.EzPySocket, data: dict):
    """[summary] The actual function that will be looped while streaming

    Args:
        c (ps.EzPySocket): [Socket object]
        data (dict): [Collection of all data needs to be passed through, this
        can be a empty dict too]
    """
    # Collect data
    ret, frame = data["camera"].read()
    if ret:
        frame = cv2.resize(frame, (1920, 1080))

        # Send image to server
        c.send_image(frame)

        # Receive image from server
        frame = c.receive_image()

        cv2.imshow("frame", frame)
        if cv2.waitKey(1) == ord('q'):
            c.stop_loop()


# Modes of operation
# 0 - Server is up until Client has gotten its request
MODE = 0

if MODE == 0:
    # Toggle to compare FPS when processing is done on server vs within script
    WITH_SERVER = True
    # Init
    cam = cv2.VideoCapture(0)
    print("Press key 'q' to exit !")
    c = None
    if WITH_SERVER:
        c = ps.EzPySocket(tokens=["start", "end"],
                          reconnect_on_address_busy=5.0,
                          server_mode=False)
        # Set configuration as needed (Increase this value incase end token is
        # not being read, i.e. give time for packets to be written properly from
        # server end
        # c.set_sleep_between_packets(0.00005)

    data = {"camera": cam}
    if WITH_SERVER:
        c.client_loop(client_operation, data)

    cv2.destroyAllWindows()
    cam.release()

File: cpp/examples/3.Webcam/run_client.cpp
#include "ezcppsocket.h"
#include <opencv2/imgproc/imgproc.hpp>

int MODE = 0;

void client_operation(EzCppSocket &c){
    static cv::VideoCapture cap(0);
    static cv::Mat frame, result;
    static bool first_run = true;
    if (first_run){
        cv::namedWindow("frame", cv::WINDOW_AUTOSIZE);
        first_run = false;
    }

    if (!cap.isOpened()) {
        std::cerr << "ERROR: Could not open camera\n";
        exit(1);
    }

    if (c.getLoopFlag() == false){
        cap.release();
        cv::destroyAllWindows();
    }

    cap >> frame;
    cv::resize(frame, frame, cv::Size(1920, 1080));
    if (!frame.empty())
    {
        // Send image to server
        c.sendImage(frame);

        // Receive image from server
        result = c.readImage();

        cv::imshow("frame", result);
        if (cv::waitKey(1) == 27){
            c.stopLoop();
        }
    }
}

int main(int argc, char **argv)
{
    EzCppSocket c = EzCppSocket("127.0.0.1", 10000, 2, 1, false, true, 1, false, 5, std::pair<std::string, std::string>("start", "end"));
    std::cout << "Press Esc to exit !\n";

    // Set configuration as needed (Increase this value incase end token is
    // not being read, i.e. give time for packets to be written properly from
    // server end
    c.setSleepBetweenPackets(100); 

    cv::Mat frame, result;
    if (MODE==0)
        c.clientLoop(&client_operation);

    return 0;
}